psycopg/psycopgmodule.c

```diff
@@ -70,24 +70,104 @@ HIDDEN PyObject *psyco_null = NULL;
 /* The type of the cursor.description items */
 HIDDEN PyObject *psyco_DescriptionType = NULL;
 
+
+/* finds a keyword or positional arg (pops it from kwargs if found there) */
+static PyObject *
+parse_arg(int pos, char *name, PyObject *defval, PyObject *args, PyObject *kwargs)
+{
+    Py_ssize_t nargs = PyTuple_GET_SIZE(args);
+    PyObject *val = NULL;
+
+    if (kwargs && PyMapping_HasKeyString(kwargs, name)) {
+        val = PyMapping_GetItemString(kwargs, name);
+        Py_XINCREF(val);
+        PyMapping_DelItemString(kwargs, name); /* pop from the kwargs dict! */
+    }
+    if (nargs > pos) {
+        if (!val) {
+            val = PyTuple_GET_ITEM(args, pos);
+            Py_XINCREF(val);
+        } else {
+            PyErr_Format(PyExc_TypeError,
+                         "parse_args() got multiple values for keyword argument '%s'", name);
+            return NULL;
+        }
+    }
+    if (!val) {
+        val = defval;
+        Py_XINCREF(val);
+    }
+
+    return val;
+}
+
+
+#define psyco_parse_args_doc \
+"parse_args(...) -- parse connection parameters.\n\n" \
+"Return a tuple of (dsn, connection_factory, async)"
+
+PyObject *
+psyco_parse_args(PyObject *self, PyObject *args, PyObject *kwargs)
+{
+    Py_ssize_t nargs = PyTuple_GET_SIZE(args);
+    PyObject *dsn = NULL;
+    PyObject *factory = NULL;
+    PyObject *async = NULL;
+    PyObject *res = NULL;
+
+    if (nargs > 3) {
+        PyErr_Format(PyExc_TypeError,
+                     "parse_args() takes at most 3 arguments (%d given)", (int)nargs);
+        goto exit;
```

Comment:
The thing is you still need to do the cleanup, and you don't know which of the objects were initialized, so there still would be some XDECREFS on NULLs.  They don't hurt anyway.

The use of goto is an issue of 1) personal taste and 2) context.  For one, I don't have problems with that (as long as it's forward goto for cleanup), and 2) there's plenty of harmless gotos around in this code base. :-)